rasa_core/policies/torch_policy.py

```diff
@@ -0,0 +1,241 @@
+        val_y = torch.tensor(shuffled_y[train_size:]).type(torch.float)
+
+        del shuffled_X, shuffled_y
+
+        logger.info("Fitting model with {} total samples and validation split "
+                    "of {}".format(training_data.num_examples(), val_split))
+
+        # hyperparams
+        epochs = kwargs.get("epochs", 30)
+        batch_size = kwargs.get("batch_size", 16)
+        # loss_fn = nn.NLLLoss(ignore_index=-1,size_average=True)
+        for ep in range(epochs):
+            permutation = torch.randperm(train_X.shape[0])
+            print_loss = 0
+            train_acc = 0
+            print_val_loss = 0
+            val_acc = 0
+            train_it = 0
+            val_it = 0
+
+            # train batches
+            for i in range(0, train_X.shape[0], batch_size):
+                optimizer.zero_grad()
+                idx = permutation[i:i + batch_size]
+                out = self.model(train_X[idx])[:, -1, :]
+                loss = -torch.mean(torch.sum(train_y[idx] * out, dim=1))
+                pred_y = torch.argmax(out, dim=1)
+                train_acc += accuracy_score(torch.argmax(train_y[idx], dim=1),
+                                            pred_y)
+                print_loss += loss.item()
+                train_it += 1
+                loss.backward()
+                optimizer.step()
+
+            for i in range(0, val_X.shape[0], batch_size):
+                val_it += 1
+                with torch.no_grad():
+                    out = self.model(val_X[i:i + batch_size])[:, -1, :]
+                    loss = -torch.mean(torch.sum(val_y[i:i + batch_size]
+                                                 * out, dim=1))
+                    pred_y = torch.argmax(out, dim=1)
+                    val_acc += accuracy_score(
+                        torch.argmax(val_y[i:i + batch_size], dim=1),
+                        pred_y)
+
+                    print_val_loss += loss.item()
+
+            logger.info("Epoch {} | Train loss/acc: {} / {} | "
+                        "Val loss/acc {} / {}"
+                        .format(ep + 1,
+                                round(print_loss, 3),
+                                round(train_acc / train_it, 3),
+                                round(print_val_loss, 3),
+                                round(val_acc / val_it, 3)))
+            self.current_epoch += 1
+
+    # TODO: test this function
+    def _continue_training(self, training_trackers, domain, **kwargs):
+        logger.info("Continuing training for model")
+        batch_size = kwargs.get("batch_size", 16)
+        epochs = kwargs.get("epochs", 10)
+
+        # TODO: use same optimizer from train()
+        optimizer = nn.optim.Adam(self.model.parameters(), lr=1e-3)
+        for ep in range(epochs):
+            training_data = self._training_data_for_continue_training(
+                batch_size, training_trackers, domain)
+            X, y = _convert_to_tensors(training_data)
```

Comment:
check the use of `_convert_to_tensors()` though it is not being used during the training process, I tried testing it and the code breaks at this point.

Reply:
This is from `_continue_training()` which is optional to implement for `Policy`. This feature will be something to consider later on.

Reply:
resolved in latest commit